src/analysis.zig

```diff
@@ -1853,13 +1859,29 @@ fn resolveTypeOfNodeUncached(analyser: *Analyser, node_handle: NodeWithHandle) e
                 label_token: ?Ast.TokenIndex,
                 then_expr: Ast.Node.Index,
                 else_expr: Ast.Node.Index,
-            } = if (ast.fullWhile(tree, node)) |while_node|
-                .{
+            } = if (ast.fullWhile(tree, node)) |while_node| no_hang: {
+
+                // while (true) {} is a noreturn type
```

Comment:
For this to be correct, you will need to check that the loop has no `break`s. Something like `while (true) { break 5; };` resolves to `comptime_int` instead of `noreturn`. 
We do have the code for resolving the type of `for` or `while` loops based on all the break operand but it is very incomplete. You could work on improving it but that is not really related to the `orelse` operation. Just like you mentioned, an implementation of peer type resolution would be needed.

Reply:
Yeah, this was a bit premature. Let's wait for the proper implementation.